In [1]:
#필요한 import
import pandas as pd
import numpy as np

# 과제1. Wikipedia에서 테이블을 가진 페이지에서 DF추출해보기.
>(어떤 페이지든 좋습니다. DF 형태가 된다면.)

Wikipedia의 테이블을 가진 페이지에서 DF를 추출하기 위해 아이유의 음반 목록 페이지를 선택했습니다.

In [2]:
url = 'https://ko.wikipedia.org/wiki/%EC%95%84%EC%9D%B4%EC%9C%A0%EC%9D%98_%EC%9D%8C%EB%B0%98_%EB%AA%A9%EB%A1%9D'
dfs = pd.read_html(url)
len(dfs)

14

dfs = pd.read_html를 통해 Wikipedia의 링크를 가져왔습니다.

len(dfs)를 통해 url에서 테이블이 14개 있다는 것을 알 수 있습니다.

In [3]:
dfs = pd.read_html(url)
dfs[9]

연도                       타이틀곡  \
                           연도                       타이틀곡   
                           연도                       타이틀곡   
0                        2008                         미아   
1                        2009                        Boo   
2                        2009                      마쉬멜로우   
3                        2010                        잔소리   
4                        2010                       좋은 날   
5                        2011                 나만 몰랐던 이야기   
6                        2011                       너랑 나   
7                        2012                       하루 끝   
8                        2013                        분홍신   
9                        2013                   금요일에 만나요   
10                       2014                  나의 옛날 이야기   
11                       2015                        스물셋   
12                       2017                        팔레트   
13                       2017            잠 못 드는 밤 비는 내리고   
14                       2018                         삐삐   
15                       2019                   Blueming   
16  "—" 표시는 차트에 진입하지 못함을 의미함.  "—" 표시는 차트에 진입하지 못함을 의미함.   

                     대한민국(가온)                             \
                          디지털                              
                           주간                         월간   
0                          57                         63   
1                          12                         13   
2                          12                         17   
3                           1                          1   
4                           1                          1   
5                           1                          1   
6                           1                          1   
7                           1                          2   
8                           1                          1   
9                           1                          5   
10                          2                          3   
11                          1                          3   
12                          1                          4   
13                          4                         11   
14                          1                          1   
15                          1                          1   
16  "—" 표시는 차트에 진입하지 못함을 의미함.  "—" 표시는 차트에 진입하지 못함을 의미함.   

                                                          \
                                                    다운로드   
                           연간                         주간   
0                           ―                          ―   
1                          66                          ―   
2                           ―                          ―   
3                           2                          1   
4                          17                          1   
5                          10                          1   
6                          36                          1   
7                          13                          1   
8                          73                          1   
9                          10                          1   
10                         14                          1   
11                         60                          1   
12                          7                          1   
13                          ―                          3   
14                         44                          1   
15                         96                          1   
16  "—" 표시는 차트에 진입하지 못함을 의미함.  "—" 표시는 차트에 진입하지 못함을 의미함.   

                                                          \
                                                           
                           월간                         연간   
0                           ―                          ―   
1                           ―                          ―   
2                           ―                          ―   
3                           1                          6   
4   

아이유의 음반 목록 페이지의 테이블중에서 dfs[9] 로 하여 타이틀곡의 테이블을 가져왔습니다.

In [4]:
dfs = pd.read_html(url, na_values='―')
dfs[9][:-1]

연도             타이틀곡 대한민국(가온)                                         \
      연도             타이틀곡      디지털          다운로드           스트리밍             
      연도             타이틀곡       주간  월간   연간   주간   월간   연간   주간   월간   연간   
0   2008               미아       57  63  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1   2009              Boo       12  13   66  NaN  NaN  NaN  NaN  NaN  NaN   
2   2009            마쉬멜로우       12  17  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
3   2010              잔소리        1   1    2    1    1    6    1    1    3   
4   2010             좋은 날        1   1   17    1    1   37    1    1    1   
5   2011       나만 몰랐던 이야기        1   1   10    1    1   51    1    2   13   
6   2011             너랑 나        1   1   36    1    1   15    1    1   25   
7   2012             하루 끝        1   2   13    1    1   15    1    4   12   
8   2013              분홍신        1   1   73    1    1   42    1    2   35   
9   2013         금요일에 만나요        1   5   10    1    5    6    1    3    3   
10  2014        나의 옛날 이야기        2   3   14    1    2   15    1    5   24   
11  2015              스물셋        1   3   60    1    3   47    1    3   71   
12  2017              팔레트        1   4    7    1    3    4    1    2    7   
13  2017  잠 못 드는 밤 비는 내리고        4  11  NaN    3    6  NaN    4   12  NaN   
14  2018               삐삐        1   1   44    1    1   22    1    2   47   
15  2019         Blueming        1   1   96    1    2  141    1    1   95   

                         음반           인증  
                         음반           인증  
                         음반           인증  
0            Lost and Found          NaN  
1                Growing Up          NaN  
2                   IU...IM          NaN  
3                       잔소리          NaN  
4                      REAL          NaN  
5                     Real+          NaN  
6              Last Fantasy          NaN  
7                   스무 살의 봄          NaN  
8              Modern Times          NaN  
9   Modern Times - Epilogue          NaN  
10                      꽃갈피          NaN  
11               CHAT-SHIRE          NaN  
12                  Palette          NaN  
13                    꽃갈피 둘          NaN  
14                       삐삐  KMCA : 플래티넘  
15                Love poem          NaN

dfs = pd.read_html(url, na_values='―')를 통해 url에서 '―'로 되어있는 값을 Na값으로 바꿔줍니다.

dfs[9][:-1]에서 첫 번째 [ ]는 가져올 테이블을 선택해주고 두 번째 [ ]는 몇 번째 행 까지 가져올 것인지에 관한 것을 정해줍니다.
>두번째[ ]에서 [:-1]로 설정하여 마지막 행을 제외한 부분을 가져왔습니다.

# 과제2. movies.csv 파일이 메모리에서 차지하는 용량을 줄이려고 합니다. 어떤 방법들이 있고, 시행했을 때 얼마나 메모리가 줄어들까요?

In [5]:
movies = pd.read_csv('data/movie.csv', nrows=4000)
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3989 non-null   object 
 1   director_name              3917 non-null   object 
 2   num_critic_for_reviews     3986 non-null   float64
 3   duration                   3993 non-null   float64
 4   director_facebook_likes    3917 non-null   float64
 5   actor_3_facebook_likes     3992 non-null   float64
 6   actor_2_name               3996 non-null   object 
 7   actor_1_facebook_likes     4000 non-null   float64
 8   gross                      3540 non-null   float64
 9   genres                     4000 non-null   object 
 10  actor_1_name               4000 non-null   object 
 11  movie_title                4000 non-null   object 
 12  num_voted_users            4000 non-null   int64  
 13  cast_total_facebook_likes  4000 non-null   int64

우선 movie.csv 파일의 메모리가 차지하는 용량을 알아보기 위해 movies.info를 사용합니다.

info 함수를 사용하여 보면 nrow를 4000으로 설정하여 4000개의 행을 불러오는데 875.1KB의 메모리가 필요합니다.
>전체 row가 4917인데 전체로하면 1.1MB로 나와 십의 자리수가 나오지않아 KB로 계산하고 싶어 임의로 nrow값을 4000으로 설정했습니다.

In [6]:
np.finfo('float64')

finfo(resolution=1e-15, min=-1.7976931348623157e+308, max=1.7976931348623157e+308, dtype=float64)

In [7]:
np.finfo('float32')

finfo(resolution=1e-06, min=-3.4028235e+38, max=3.4028235e+38, dtype=float32)

In [8]:
np.finfo('float16')

finfo(resolution=0.001, min=-6.55040e+04, max=6.55040e+04, dtype=float16)

In [9]:
np.iinfo('int64')

iinfo(min=-9223372036854775808, max=9223372036854775807, dtype=int64)

In [10]:
np.iinfo('int32')

iinfo(min=-2147483648, max=2147483647, dtype=int32)

In [11]:
np.iinfo('int16')

iinfo(min=-32768, max=32767, dtype=int16)

In [12]:
np.iinfo('int8')

iinfo(min=-128, max=127, dtype=int8)

Dtype으로 변경하기 위해서 float과 int의 값을 찾아봤을 때 뒤의 숫자가 클수록 더 정교한 숫자 즉 더 크고 작은 수를 표현 가능 또한 뒤의 숫자가 클수록 더 많은 메모리를 차지

In [13]:
movies.head() #데이터 확인 과정

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4    NaN        Doug Walker                     NaN       NaN   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    131.0                     NaN        Rob Walker   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   131.0          NaN                      Documentary  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                  NaN      NaN      NaN             NaN          NaN   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4         NaN                   12.0        7.1           NaN   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                    0  

[5 rows x 28 columns]

In [14]:
movies.describe().T

count          mean           std       min  \
num_critic_for_reviews     3986.0  1.525354e+02  1.222510e+02      1.00   
duration                   3993.0  1.096759e+02  2.597698e+01      7.00   
director_facebook_likes    3917.0  7.777347e+02  3.015621e+03      0.00   
actor_3_facebook_likes     3992.0  7.258770e+02  1.782981e+03      0.00   
actor_1_facebook_likes     4000.0  7.262465e+03  1.471841e+04      0.00   
gross                      3540.0  5.364939e+07  6.977666e+07    162.00   
num_voted_users            4000.0  9.621417e+04  1.474648e+05      6.00   
cast_total_facebook_likes  4000.0  1.082019e+04  1.818425e+04      0.00   
facenumber_in_poster       3993.0  1.405710e+00  2.056854e+00      0.00   
num_user_for_reviews       3991.0  3.037537e+02  3.931481e+02      1.00   
budget                     3691.0  4.319484e+07  1.073322e+08  13000.00   
title_year                 3913.0  2.002909e+03  1.065295e+01   1927.00   
actor_2_facebook_likes     3996.0  1.879982e+03  4.340798e+03      0.00   
imdb_score                 4000.0  6.441375e+00  1.093554e+00      1.60   
aspect_ratio               3881.0  2.221180e+00  1.228050e+00      1.18   
movie_facebook_likes       4000.0  8.499153e+03  2.079905e+04      0.00   

                                   25%          50%          75%           max  
num_critic_for_reviews           63.00       123.00       208.00  8.130000e+02  
duration                         95.00       106.00       120.00  5.110000e+02  
director_facebook_likes          10.00        57.00       218.00  2.300000e+04  
actor_3_facebook_likes          184.00       427.50       681.00  2.300000e+04  
actor_1_facebook_likes          727.00      1000.00     12000.00  6.400000e+05  
gross                      10199691.50  31494026.00  68917924.75  7.605058e+08  
num_voted_users               14786.00     45251.00    111915.25  1.689764e+06  
cast_total_facebook_likes      1834.00      3797.50     15433.25  6.567300e+05  
facenumber_in_poster              0.00         1.00         2.00  4.300000e+01  
num_user_for_reviews             89.00       183.00       365.00  5.060000e+03  
budget                     12000000.00  25000000.00  50000000.00  4.200000e+09  
title_year                     1999.00      2005.00      2010.00  2.016000e+03  
actor_2_facebook_likes          370.00       664.00       962.00  1.370000e+05  
imdb_score                        5.80         6.50         7.20  9.500000e+00  
aspect_ratio                      1.85         2.35         2.35  1.600000e+01  
movie_facebook_likes              0.00       204.00     10000.00  3.490000e+05

대부분의 float은 float16의 값으로 만족하지만 actor_1_facebook_likes, gross, budget, actor_2_facebook_likes는 float16의 값을 넘기 때문에 float32를 사용해야합니다.

int컬럼의 최대값은 전부 int16의 최대값인 32767 이상이기 때문에 int32를 사용합니다. 

데이터를 변경할 때 손실이 일어나면 안됩니다. 지금은 row를 4000개만 가져왔지만 전체 데이터를 가져왔을 때 문제없도록 소수점 min과 max의 값을 충분하게 표현할 수 있게 여유 있게 정해줘야합니다.

## float과 int의 Dtype을 변경

In [15]:
# float과 int의 Dtype을 변경
movies2 = pd.read_csv('data/movie.csv', nrows=4000,
                     dtype={
                         'num_critic_for_reviews':np.float16,
                         'duration':np.float16,
                         'director_facebook_likes':np.float16,
                         'actor_3_facebook_likes':np.float16,
                         'actor_1_facebook_likes':np.float32,
                         'gross':np.float32,
                         'facenumber_in_poster':np.float16,
                         'num_user_for_reviews':np.float16,
                         'budget':np.float32,
                         'title_year':np.float16,
                         'actor_2_facebook_likes':np.float32,
                         'imdb_score':np.float16,
                         'aspect_ratio':np.float16,
                         'num_voted_users':np.int32,
                         'cast_total_facebook_likes':np.int32,
                         'movie_facebook_likes':np.int32,
                     })
movies2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3989 non-null   object 
 1   director_name              3917 non-null   object 
 2   num_critic_for_reviews     3986 non-null   float16
 3   duration                   3993 non-null   float16
 4   director_facebook_likes    3917 non-null   float16
 5   actor_3_facebook_likes     3992 non-null   float16
 6   actor_2_name               3996 non-null   object 
 7   actor_1_facebook_likes     4000 non-null   float32
 8   gross                      3540 non-null   float32
 9   genres                     4000 non-null   object 
 10  actor_1_name               4000 non-null   object 
 11  movie_title                4000 non-null   object 
 12  num_voted_users            4000 non-null   int32  
 13  cast_total_facebook_likes  4000 non-null   int32

In [16]:
round(554.8/875.1, 2)

0.63

처음 데이터 875.1KB에서 현재 554.8KB로 처음과 비교해 63%의 메모리만 사용합니다.

In [17]:
movies2.describe(include='O')

color     director_name    actor_2_name  genres    actor_1_name  \
count    3989              3917            3996    4000            4000   
unique      2              1793            2359     813            1557   
top     Color  Steven Spielberg  Morgan Freeman  Comedy  Robert De Niro   
freq     3871                26              18     165              47   

       movie_title  actor_3_name   plot_keywords  \
count         4000          3992            3939   
unique        4000          2793            3937   
top         Avatar  Steve Coogan  based on novel   
freq             1             8               3   

                                          movie_imdb_link language country  \
count                                                4000     3994    3995   
unique                                               4000       35      53   
top     http://www.imdb.com/title/tt0499549/?ref_=fn_t...  English     USA   
freq                                                    1     3775    3031   

       content_rating  
count            3878  
unique             18  
top                 R  
freq             1674

movies.describe(include='O')를 통해 category로 유의미하게 변할 수 있는 변수를 찾을 수 있습니다.

unique는 각 컬럼이 몇개의 고유한 값을 가지는지 나타냅니다.

category 형태로 변환할 때 특정한 값 몇개가 반복될 때 유의미한 결과가 나오는 것이지 movie_title 컬럼과 같이 unique값이 너무 많으면 변환하는 의미가 없습니다. 
> 그렇기 때문에 category 변환으로 선택한 object컬럼은 color, language, country, content_rating 입니다.

category 형태로 변환할 때 데이터가 길수록, 종류가 적을수록 공간 확보의 효과가 큽니다.

In [18]:
movies2.color.value_counts()

Color              3871
Black and White     118
Name: color, dtype: int64

In [19]:
movies2.language.value_counts()

English       3775
French          56
Spanish         24
Mandarin        22
Hindi           17
Japanese        14
German          12
Russian         11
Cantonese        9
Italian          6
Korean           6
Portuguese       5
Thai             3
Swedish          3
Chinese          3
Dutch            3
Polish           2
Norwegian        2
Danish           2
Aboriginal       2
None             2
Icelandic        2
Bosnian          1
Czech            1
Arabic           1
Hungarian        1
Aramaic          1
Filipino         1
Maya             1
Kazakh           1
Mongolian        1
Telugu           1
Hebrew           1
Dari             1
Kannada          1
Name: language, dtype: int64

In [20]:
movies2.country.value_counts()

USA               3031
UK                 368
France             129
Canada              92
Germany             82
Australia           44
Spain               27
China               26
India               20
Japan               19
Hong Kong           16
Italy               16
South Korea         12
Mexico              11
Russia              11
New Zealand          9
Ireland              8
Denmark              7
South Africa         6
Norway               5
Thailand             4
Brazil               4
Sweden               4
Netherlands          4
Iceland              3
Belgium              3
Czech Republic       3
Switzerland          2
West Germany         2
Hungary              2
Taiwan               2
Poland               2
Libya                1
Bulgaria             1
Argentina            1
Bahamas              1
Finland              1
Nigeria              1
Turkey               1
Georgia              1
Iran                 1
Cambodia             1
Slovakia             1
New Line   

In [21]:
movies2.content_rating.value_counts()

R            1674
PG-13        1307
PG            620
G             102
Not Rated      41
Approved       27
Unrated        25
TV-14          25
TV-MA          16
TV-PG          10
TV-G            9
X               9
GP              3
Passed          3
M               3
NC-17           2
TV-Y            1
TV-Y7           1
Name: content_rating, dtype: int64

## float과 int와 object를category로의 Dtype을 변경

In [22]:
movies3 = pd.read_csv('data/movie.csv', nrows=4000,
                     dtype={
                         'num_critic_for_reviews':np.float16,
                         'duration':np.float16,
                         'director_facebook_likes':np.float16,
                         'actor_3_facebook_likes':np.float16,
                         'actor_1_facebook_likes':np.float32,
                         'gross':np.float32,
                         'facenumber_in_poster':np.float16,
                         'num_user_for_reviews':np.float16,
                         'budget':np.float32,
                         'title_year':np.float16,
                         'actor_2_facebook_likes':np.float32,
                         'imdb_score':np.float16,
                         'aspect_ratio':np.float16,
                         'num_voted_users':np.int32,
                         'cast_total_facebook_likes':np.int32,
                         'movie_facebook_likes':np.int32,
                         'color':'category',
                         'language':'category',
                         'country':'category',
                         'content_rating':'category',
                     })
movies3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype   
---  ------                     --------------  -----   
 0   color                      3989 non-null   category
 1   director_name              3917 non-null   object  
 2   num_critic_for_reviews     3986 non-null   float16 
 3   duration                   3993 non-null   float16 
 4   director_facebook_likes    3917 non-null   float16 
 5   actor_3_facebook_likes     3992 non-null   float16 
 6   actor_2_name               3996 non-null   object  
 7   actor_1_facebook_likes     4000 non-null   float32 
 8   gross                      3540 non-null   float32 
 9   genres                     4000 non-null   object  
 10  actor_1_name               4000 non-null   object  
 11  movie_title                4000 non-null   object  
 12  num_voted_users            4000 non-null   int32   
 13  cast_total_facebook_likes  4000 n

In [23]:
round(450.0/875.1, 2)

0.51

처음 데이터 875.1KB에서 현재 450.0KB로 처음과 비교해 51%의 메모리만 사용합니다.

## usecols=cols를 사용하여 특정 컬럼만 선택

위와 같은 Dtype변경 후에도 메모리가 모자라면 특정 컬럼만 선택하는 방법이 있습니다.

특정 컬럼을 선택해서 내가 필요없는 부분을 제외하고 불러오는 방법인

read.csv의 usecols를 사용해서 cols에 내가 원하는 컬럼들의 이름을 정해주면 그 컬럼들만 선택해서 가져옵니다.

In [24]:
#cols에서  'movie_imdb_link', 'plot_keywords'을 제외
cols = ['color', 'director_name', 'num_critic_for_reviews', 'duration', 'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name', 
        'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name', 'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
        'actor_3_name', 'facenumber_in_poster', 'num_user_for_reviews', 'language', 'country',
        'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes', 'imdb_score', 'aspect_ratio', 'movie_facebook_likes']

movies4 = pd.read_csv('data/movie.csv', nrows=4000,
                     dtype={
                         'num_critic_for_reviews':np.float16,
                         'duration':np.float16,
                         'director_facebook_likes':np.float16,
                         'actor_3_facebook_likes':np.float16,
                         'actor_1_facebook_likes':np.float32,
                         'gross':np.float32,
                         'facenumber_in_poster':np.float16,
                         'num_user_for_reviews':np.float16,
                         'budget':np.float32,
                         'title_year':np.float16,
                         'actor_2_facebook_likes':np.float32,
                         'imdb_score':np.float16,
                         'aspect_ratio':np.float16,
                         'num_voted_users':np.int32,
                         'cast_total_facebook_likes':np.int32,
                         'movie_facebook_likes':np.int32,
                         'color':'category',
                         'language':'category',
                         'country':'category',
                         'content_rating':'category',
                     },
                     usecols=cols)
movies4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype   
---  ------                     --------------  -----   
 0   color                      3989 non-null   category
 1   director_name              3917 non-null   object  
 2   num_critic_for_reviews     3986 non-null   float16 
 3   duration                   3993 non-null   float16 
 4   director_facebook_likes    3917 non-null   float16 
 5   actor_3_facebook_likes     3992 non-null   float16 
 6   actor_2_name               3996 non-null   object  
 7   actor_1_facebook_likes     4000 non-null   float32 
 8   gross                      3540 non-null   float32 
 9   genres                     4000 non-null   object  
 10  actor_1_name               4000 non-null   object  
 11  movie_title                4000 non-null   object  
 12  num_voted_users            4000 non-null   int32   
 13  cast_total_facebook_likes  4000 n

In [25]:
round(387.5/875.1, 2)

0.44

cols에서 임의로 'movie_imdb_link', 'plot_keywords'컬럼을 제외한 경우 처음 데이터 875.1KB에서 현재 387.5KB로 처음과 비교해 44%의 메모리만 사용합니다.

## chunksize 사용

만약 이렇게 했는데도 공간이 모자라면 데이터를 나눠서 로딩하는 방법이 있습니다.

read_csv 에서 chunksize를 설정해 데이터를 한번에 몇 줄을 불러올 것인지를 정합니다. 

In [26]:
cols = ['color', 'director_name', 'num_critic_for_reviews', 'duration', 'director_facebook_likes', 'actor_3_facebook_likes', 'actor_2_name', 
        'actor_1_facebook_likes', 'gross', 'genres', 'actor_1_name', 'movie_title', 'num_voted_users', 'cast_total_facebook_likes',
        'actor_3_name', 'facenumber_in_poster', 'num_user_for_reviews', 'language', 'country',
        'content_rating', 'budget', 'title_year', 'actor_2_facebook_likes', 'imdb_score', 'aspect_ratio', 'movie_facebook_likes']

movies_iter = pd.read_csv('data/movie.csv', nrows=4000,
                     dtype={
                         'num_critic_for_reviews':np.float16,
                         'duration':np.float16,
                         'director_facebook_likes':np.float16,
                         'actor_3_facebook_likes':np.float16,
                         'actor_1_facebook_likes':np.float32,
                         'gross':np.float32,
                         'facenumber_in_poster':np.float16,
                         'num_user_for_reviews':np.float16,
                         'budget':np.float32,
                         'title_year':np.float16,
                         'actor_2_facebook_likes':np.float32,
                         'imdb_score':np.float16,
                         'aspect_ratio':np.float16,
                         'num_voted_users':np.int32,
                         'cast_total_facebook_likes':np.int32,
                         'movie_facebook_likes':np.int32,
                         'color':'category',
                         'language':'category',
                         'country':'category',
                         'content_rating':'category',
                     },
                     usecols=cols,
                     chunksize=500)

def process(df):
    return f'processed {df.shape} shape of df' #{df.iloc[0, :]}

for chunk in movies_iter:
    print(process(chunk))

processed (500, 26) shape of df
processed (500, 26) shape of df
processed (500, 26) shape of df
processed (500, 26) shape of df
processed (500, 26) shape of df
processed (500, 26) shape of df
processed (500, 26) shape of df
processed (500, 26) shape of df


4000개의 row를 500개씩 나눠서 불러온 것을 알 수 있습니다.